diff --git inline-c-cpp/inline-c-cpp.cabal inline-c-cpp/inline-c-cpp.cabal
index 7985158..20f5468 100644
--- inline-c-cpp/inline-c-cpp.cabal
+++ inline-c-cpp/inline-c-cpp.cabal
@@ -1,3 +1,4 @@
+cabal-version:       2.4
 name:                inline-c-cpp
 version:             0.3.0.2
 synopsis:            Lets you embed C++ code into Haskell.
@@ -11,7 +12,6 @@ copyright:           (c) 2015-2016 FP Complete Corporation, (c) 2017-2019 France
 category:            FFI
 tested-with:         GHC == 8.2.2, GHC == 8.4.4, GHC == 8.6.5
 build-type:          Simple
-cabal-version:       >=1.10
 
 source-repository head
   type:     git
@@ -26,7 +26,11 @@ library
                      , safe-exceptions
   hs-source-dirs:      src
   default-language:    Haskell2010
-  ghc-options:         -Wall -optc-std=c++11
+  cxx-options:         --std=c++17
+  ghc-options:         -Wall
+  extra-libraries:     stdc++
+  if os(darwin)
+    ghc-options: -pgmc=clang++
   if os(darwin)
     -- avoid https://gitlab.haskell.org/ghc/ghc/issues/11829
     ld-options:        -Wl,-keep_dwarf_unwind
@@ -42,9 +46,3 @@ test-suite tests
                      , safe-exceptions
                      , hspec
   default-language:    Haskell2010
-  ghc-options:
-    -optc-std=c++11
-  if os(darwin)
-    ghc-options: -pgmc=clang++
-  extra-libraries:     stdc++
-  cc-options:          -Wall -Werror -std=c++11

